File: src/retro_metadata/providers/steamgriddb.py
"""SteamGridDB artwork provider implementation."""

from __future__ import annotations

import json
import logging
import re
from enum import StrEnum
from typing import TYPE_CHECKING, Any, Final

import httpx

from retro_metadata.core.exceptions import (
    ProviderAuthenticationError,
    ProviderConnectionError,
    ProviderRateLimitError,
)
from retro_metadata.providers.base import MetadataProvider
from retro_metadata.types.common import (
    Artwork,
    GameMetadata,
    GameResult,
    SearchResult,
)

if TYPE_CHECKING:
    from retro_metadata.cache.base import CacheBackend
    from retro_metadata.core.config import ProviderConfig

logger = logging.getLogger(__name__)


class SGDBDimension(StrEnum):
    """SteamGridDB grid dimension options."""

    # Vertical grids
    STEAM_VERTICAL = "600x900"
    GOG_GALAXY = "342x482"
    SQUARE = "512x512"
    SQUARE_ICON = "256x256"

    # Horizontal grids
    STEAM_HORIZONTAL = "460x215"
    LEGACY = "920x430"
    OLD = "460x215"

    # Heroes
    HERO_BLURRED = "1920x620"
    HERO_MATERIAL = "3840x1240"

    # Logos
    LOGO_CUSTOM = "custom"


class SGDBStyle(StrEnum):
    """SteamGridDB artwork style options."""

    # Grid styles
    ALTERNATE = "alternate"
    BLURRED = "blurred"
    WHITE_LOGO = "white_logo"
    MATERIAL = "material"
    NO_LOGO = "no_logo"

    # Hero styles
    HERO_ALTERNATE = "alternate"
    HERO_BLURRED = "blurred"
    HERO_MATERIAL = "material"

    # Logo styles
    LOGO_OFFICIAL = "official"
    LOGO_WHITE = "white"
    LOGO_BLACK = "black"
    LOGO_CUSTOM = "custom"


class SGDBMime(StrEnum):
    """SteamGridDB MIME type options."""

    PNG = "image/png"
    JPEG = "image/jpeg"
    WEBP = "image/webp"
    ICO = "image/vnd.microsoft.icon"

# Regex to detect SteamGridDB ID tags in filenames like (sgdb-12345)
SGDB_TAG_REGEX: Final = re.compile(r"\(sgdb-(\d+)\)", re.IGNORECASE)


class SteamGridDBProvider(MetadataProvider):
    """SteamGridDB artwork provider.

    Provides artwork (covers, banners, logos, icons, heroes) for games.
    Note: This provider focuses on artwork rather than full metadata.

    Requires an api_key credential from SteamGridDB.

    Example:
        config = ProviderConfig(
            enabled=True,
            credentials={"api_key": "your_api_key"}
        )
        provider = SteamGridDBProvider(config)
        results = await provider.search("Super Mario World")
    """

    name = "steamgriddb"

    def __init__(
        self,
        config: ProviderConfig,
        cache: CacheBackend | None = None,
        user_agent: str = "retro-metadata/1.0",
    ) -> None:
        super().__init__(config, cache)
        self._base_url = "https://www.steamgriddb.com/api/v2"
        self._user_agent = user_agent
        self._client: httpx.AsyncClient | None = None
        # High similarity threshold for better matches (RomM uses 0.98)
        self._min_similarity_score = 0.98
        # Content filter defaults (can be configured)
        self._nsfw = config.settings.get("nsfw", False)
        self._humor = config.settings.get("humor", True)
        self._epilepsy = config.settings.get("epilepsy", True)

    @property
    def api_key(self) -> str:
        return self.config.get_credential("api_key")

    async def _get_client(self) -> httpx.AsyncClient:
        """Get or create the httpx client."""
        if self._client is None or self._client.is_closed:
            self._client = httpx.AsyncClient(
                base_url=self._base_url,
                headers={
                    "User-Agent": self._user_agent,
                    "Authorization": f"Bearer {self.api_key}",
                },
                timeout=self.config.timeout,
            )
        return self._client

    async def _request(
        self,
        endpoint: str,
        params: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Make an API request to SteamGridDB."""
        client = await self._get_client()

        logger.debug("SteamGridDB API: GET %s%s", self._base_url, endpoint)
        if params:
            logger.debug("SteamGridDB API params: %s", params)

        try:
            response = await client.get(endpoint, params=params)

            if response.status_code == 401:
                logger.debug("SteamGridDB API: 401 Unauthorized")
                raise ProviderAuthenticationError(self.name, "Invalid API key")
            elif response.status_code == 429:
                logger.debug("SteamGridDB API: 429 Rate limited")
                raise ProviderRateLimitError(self.name)

            response.raise_for_status()
            data = response.json()

            # Log full response body only when debug logging is enabled
            if logger.isEnabledFor(logging.DEBUG):
                logger.debug("SteamGridDB API response:\n%s", json.dumps(data, indent=2, ensure_ascii=False))

            return data
        except httpx.RequestError as e:
            logger.debug("SteamGridDB API error: %s", e)
            raise ProviderConnectionError(self.name, str(e)) from e

    async def search(
        self,
        query: str,
        platform_id: int | None = None,  # noqa: ARG002
        limit: int = 10,
    ) -> list[SearchResult]:
        """Search for games by name.

        Args:
            query: Search query string
            platform_id: Not used by SteamGridDB
            limit: Maximum number of results

        Returns:
            List of search results
        """
        if not self.is_enabled:
            return []

        result = await self._request("/search/autocomplete/" + query)

        if not result.get("success") or "data" not in result:
            return []

        search_results = []
        for game in result["data"][:limit]:
            # Get cover image for the game
            cover_url = ""
            try:
                grids = await self._request(f"/grids/game/{game['id']}")
                if grids.get("success") and grids.get("data"):
                    cover_url = grids["data"][0].get("url", "")
            except Exception:
                pass

            search_results.append(
                SearchResult(
                    name=game.get("name", ""),
                    provider=self.name,
                    provider_id=game["id"],
                    cover_url=cover_url,
                    release_year=game.get("release_date"),
                )
            )

        return search_results

    async def get_by_id(self, game_id: int) -> GameResult | None:
        """Get game artwork by SteamGridDB ID.

        Args:
            game_id: SteamGridDB game ID

        Returns:
            GameResult with artwork, or None if not found
        """
        if not self.is_enabled:
            return None

        result = await self._request(f"/games/id/{game_id}")

        if not result.get("success") or "data" not in result:
            return None

        game = result["data"]

        # Fetch all artwork types
        artwork = await self._fetch_all_artwork(game_id)

        return GameResult(
            name=game.get("name", ""),
            provider=self.name,
            provider_id=game["id"],
            provider_ids={"steamgriddb": game["id"]},
            artwork=artwork,
            metadata=GameMetadata(
                release_year=game.get("release_date"),
            ),
            raw_response=game,
        )

    async def _fetch_all_artwork(self, game_id: int) -> Artwork:
        """Fetch all artwork types for a game."""
        artwork = Artwork()

        # Fetch grids (covers) with content filters
        try:
            grids = await self._fetch_grids(game_id)
            if grids:
                artwork.cover_url = grids[0].get("url", "")
        except Exception:
            pass

        # Fetch heroes (banners/backgrounds)
        try:
            heroes = await self._fetch_heroes(game_id)
            if heroes:
                artwork.background_url = heroes[0].get("url", "")
                if len(heroes) > 1:
                    artwork.banner_url = heroes[1].get("url", "")
        except Exception:
            pass

        # Fetch logos
        try:
            logos = await self._fetch_logos(game_id)
            if logos:
                artwork.logo_url = logos[0].get("url", "")
        except Exception:
            pass

        # Fetch icons
        try:
            icons = await self._fetch_icons(game_id)
            if icons:
                artwork.icon_url = icons[0].get("url", "")
        except Exception:
            pass

        return artwork

    def _build_filter_params(
        self,
        dimensions: list[SGDBDimension] | None = None,
        styles: list[SGDBStyle] | None = None,
        mimes: list[SGDBMime] | None = None,
        types: list[str] | None = None,
    ) -> dict[str, str]:
        """Build filter parameters for artwork requests."""
        params: dict[str, str] = {}

        # Content filters
        nsfw = "any" if self._nsfw else "false"
        humor = "any" if self._humor else "false"
        epilepsy = "any" if self._epilepsy else "false"

        params["nsfw"] = nsfw
        params["humor"] = humor
        params["epilepsy"] = epilepsy

        if dimensions:
            params["dimensions"] = ",".join(dimensions)
        if styles:
            params["styles"] = ",".join(styles)
        if mimes:
            params["mimes"] = ",".join(mimes)
        if types:
            params["types"] = ",".join(types)

        return params

    async def _fetch_grids(
        self,
        game_id: int,
        dimensions: list[SGDBDimension] | None = None,
        styles: list[SGDBStyle] | None = None,
    ) -> list[dict[str, Any]]:
        """Fetch grid artwork with optional filters."""
        params = self._build_filter_params(dimensions=dimensions, styles=styles)
        result = await self._request(f"/grids/game/{game_id}", params or None)
        if result.get("success") and result.get("data"):
            return result["data"]
        return []

    async def _fetch_heroes(
        self,
        game_id: int,
        dimensions: list[SGDBDimension] | None = None,
        styles: list[SGDBStyle] | None = None,
    ) -> list[dict[str, Any]]:
        """Fetch hero artwork with optional filters."""
        params = self._build_filter_params(dimensions=dimensions, styles=styles)
        result = await self._request(f"/heroes/game/{game_id}", params or None)
        if result.get("success") and result.get("data"):
            return result["data"]
        return []

    async def _fetch_logos(
        self,
        game_id: int,
        styles: list[SGDBStyle] | None = None,
    ) -> list[dict[str, Any]]:
        """Fetch logo artwork with optional filters."""
        params = self._build_filter_params(styles=styles)
        result = await self._request(f"/logos/game/{game_id}", params or None)
        if result.get("success") and result.get("data"):
            return result["data"]
        return []

    async def _fetch_icons(
        self,
        game_id: int,
        dimensions: list[SGDBDimension] | None = None,
    ) -> list[dict[str, Any]]:
        """Fetch icon artwork with optional filters."""
        params = self._build_filter_params(dimensions=dimensions)
        result = await self._request(f"/icons/game/{game_id}", params or None)
        if result.get("success") and result.get("data"):
            return result["data"]
        return []

    async def identify(
        self,
        filename: str,
        platform_id: int | None = None,  # noqa: ARG002
    ) -> GameResult | None:
        """Identify a game from a ROM filename.

        Args:
            filename: ROM filename
            platform_id: Not used by SteamGridDB

        Returns:
            GameResult if a match is found, None otherwise
        """
        if not self.is_enabled:
            return None

        # Check for SteamGridDB ID tag in filename
        tagged_id = self.extract_id_from_filename(filename, SGDB_TAG_REGEX)
        if tagged_id:
            result = await self.get_by_id(tagged_id)
            if result:
                return result

        # Clean the filename
        search_term = self._clean_filename(filename)

        # Search for the game
        result = await self._request("/search/autocomplete/" + search_term)

        if not result.get("success") or not result.get("data"):
            # Try splitting by special characters
            terms = self.split_search_term(search_term)
            if len(terms) > 1:
                result = await self._request("/search/autocomplete/" + terms[-1])

        if not result.get("success") or not result.get("data"):
            return None

        # Find best match
        games_by_name = {g["name"]: g for g in result["data"]}
        best_match, score = self.find_best_match(
            search_term, list(games_by_name.keys())
        )

        if best_match and best_match in games_by_name:
            game = games_by_name[best_match]
            artwork = await self._fetch_all_artwork(game["id"])

            game_result = GameResult(
                name=game.get("name", ""),
                provider=self.name,
                provider_id=game["id"],
                provider_ids={"steamgriddb": game["id"]},
                artwork=artwork,
                metadata=GameMetadata(
                    release_year=game.get("release_date"),
                ),
                match_score=score,
                raw_response=game,
            )
            return game_result

        return None

    def _clean_filename(self, filename: str) -> str:
        """Remove tags and extension from filename."""
        name = re.sub(r"\.[^.]+$", "", filename)
        name = re.sub(r"\s*[\(\[][^\)\]]*[\)\]]", "", name)
        return name.strip()

    async def get_artwork_for_steam_id(self, steam_app_id: int) -> Artwork:
        """Get artwork using a Steam App ID.

        Args:
            steam_app_id: Steam application ID

        Returns:
            Artwork object with available images
        """
        if not self.is_enabled:
            return Artwork()

        # Look up game by Steam ID
        result = await self._request(f"/games/steam/{steam_app_id}")

        if not result.get("success") or "data" not in result:
            return Artwork()

        game_id = result["data"]["id"]
        return await self._fetch_all_artwork(game_id)

    async def close(self) -> None:
        """Close the httpx client."""
        if self._client is not None and not self._client.is_closed:
            await self._client.aclose()

File: cpp/src/provider/steamgriddb.cpp
/// @file steamgriddb.cpp
/// @brief SteamGridDB provider implementation for artwork fetching

#include <retro_metadata/cache/cache.hpp>
#include <retro_metadata/config.hpp>
#include <retro_metadata/errors.hpp>
#include <retro_metadata/internal/matching.hpp>
#include <retro_metadata/internal/normalization.hpp>
#include <retro_metadata/provider/provider.hpp>
#include <retro_metadata/provider/registry.hpp>

#include <cpr/cpr.h>
#include <nlohmann/json.hpp>

#include <regex>
#include <string>
#include <vector>

namespace retro_metadata {

namespace {

/// Base URL for the SteamGridDB API
const std::string kBaseURL = "https://www.steamgriddb.com/api/v2";

/// Regex to detect SteamGridDB ID tags in filenames like (sgdb-12345)
const std::regex kSgdbTagRegex(R"(\(sgdb-(\d+)\))", std::regex::icase);

/// @brief SteamGridDB grid dimension options
enum class SGDBDimension {
    // Vertical grids
    SteamVertical,   // 600x900
    GOGGalaxy,       // 342x482
    Square,          // 512x512
    SquareIcon,      // 256x256
    // Horizontal grids
    SteamHorizontal, // 460x215
    Legacy,          // 920x430
    Old,             // 460x215
    // Heroes
    HeroBlurred,     // 1920x620
    HeroMaterial,    // 3840x1240
    // Logos
    LogoCustom       // custom
};

/// @brief SteamGridDB artwork style options
enum class SGDBStyle {
    // Grid styles
    Alternate,
    Blurred,
    WhiteLogo,
    Material,
    NoLogo,
    // Logo styles
    LogoOfficial,
    LogoWhite,
    LogoBlack,
    LogoCustom
};

/// @brief SteamGridDB MIME type options
enum class SGDBMime {
    PNG,
    JPEG,
    WEBP,
    ICO
};

std::string get_string(const nlohmann::json& j, const std::string& key) {
    if (j.contains(key) && j[key].is_string()) {
        return j[key].get<std::string>();
    }
    return "";
}

double get_number(const nlohmann::json& j, const std::string& key) {
    if (j.contains(key) && j[key].is_number()) {
        return j[key].get<double>();
    }
    return 0.0;
}

bool get_bool(const nlohmann::json& j, const std::string& key) {
    if (j.contains(key) && j[key].is_boolean()) {
        return j[key].get<bool>();
    }
    return false;
}

std::string clean_filename(const std::string& filename) {
    static const std::regex ext_pattern(R"(\.[^.]+$)");
    static const std::regex tag_pattern(R"(\s*[\(\[][^\)\]]*[\)\]])");
    std::string name = std::regex_replace(filename, ext_pattern, "");
    name = std::regex_replace(name, tag_pattern, "");
    auto start = name.find_first_not_of(" \t\n\r");
    if (start == std::string::npos) return "";
    auto end = name.find_last_not_of(" \t\n\r");
    return name.substr(start, end - start + 1);
}

std::string url_encode(const std::string& str) {
    std::string encoded;
    for (char c : str) {
        if (std::isalnum(static_cast<unsigned char>(c)) || c == '-' || c == '_' || c == '.' || c == '~') {
            encoded += c;
        } else if (c == ' ') {
            encoded += "%20";
        } else {
            char buf[4];
            snprintf(buf, sizeof(buf), "%%%02X", static_cast<unsigned char>(c));
            encoded += buf;
        }
    }
    return encoded;
}

}  // namespace

/// @brief SteamGridDB metadata provider for artwork
class SteamGridDBProvider : public Provider {
public:
    SteamGridDBProvider(const ProviderConfig& config, std::shared_ptr<Cache> cache)
        : config_(config), cache_(std::move(cache)), nsfw_(false), humor_(true), epilepsy_(true) {
        // Check options for content filters
        if (auto it = config_.options.find("nsfw"); it != config_.options.end()) {
            try {
                nsfw_ = std::any_cast<bool>(it->second);
            } catch (...) {}
        }
        if (auto it = config_.options.find("humor"); it != config_.options.end()) {
            try {
                humor_ = std::any_cast<bool>(it->second);
            } catch (...) {}
        }
        if (auto it = config_.options.find("epilepsy"); it != config_.options.end()) {
            try {
                epilepsy_ = std::any_cast<bool>(it->second);
            } catch (...) {}
        }
    }

    std::string name() const override { return "steamgriddb"; }

    std::vector<SearchResult> search(const std::string& query, const SearchOptions& opts) override {
        if (!config_.is_configured()) return {};

        auto response = request("/search/autocomplete/" + url_encode(query));
        if (response.empty() || !get_bool(response, "success")) {
            return {};
        }

        if (!response.contains("data") || !response["data"].is_array()) {
            return {};
        }

        int limit = opts.limit > 0 ? opts.limit : 10;
        std::vector<SearchResult> results;

        for (const auto& item : response["data"]) {
            if (static_cast<int>(results.size()) >= limit) break;

            int game_id = static_cast<int>(get_number(item, "id"));
            if (game_id == 0) continue;

            SearchResult sr;
            sr.provider = name();
            sr.provider_id = game_id;
            sr.name = get_string(item, "name");

            // Try to get cover image from grids
            auto grids = fetch_grids(game_id);
            if (!grids.empty() && grids[0].contains("url")) {
                sr.cover_url = get_string(grids[0], "url");
            }

            double release_date = get_number(item, "release_date");
            if (release_date > 0) {
                sr.release_year = static_cast<int>(release_date);
            }

            results.push_back(std::move(sr));
        }

        return results;
    }

    std::unique_ptr<GameResult> get_by_id(int game_id) override {
        if (!config_.is_configured()) return nullptr;

        auto response = request("/games/id/" + std::to_string(game_id));
        if (response.empty() || !get_bool(response, "success")) {
            return nullptr;
        }

        if (!response.contains("data") || !response["data"].is_object()) {
            return nullptr;
        }

        const auto& game = response["data"];
        auto artwork = fetch_all_artwork(game_id);

        auto result = std::make_unique<GameResult>();
        result->provider = name();
        result->provider_id = game_id;
        result->provider_ids = {{"steamgriddb", game_id}};
        result->name = get_string(game, "name");
        result->artwork = artwork;

        double release_date = get_number(game, "release_date");
        if (release_date > 0) {
            result->metadata.release_year = static_cast<int>(release_date);
        }

        result->raw_response = game;
        return result;
    }

    std::unique_ptr<GameResult> identify(const std::string& filename, const IdentifyOptions& /*opts*/) override {
        if (!config_.is_configured()) return nullptr;

        // Check for SteamGridDB ID tag in filename
        std::smatch match;
        if (std::regex_search(filename, match, kSgdbTagRegex) && match.size() > 1) {
            try {
                int id = std::stoi(match[1].str());
                auto result = get_by_id(id);
                if (result) {
                    result->match_type = "tag";
                    result->match_score = 1.0;
                    return result;
                }
            } catch (...) {}
        }

        // Clean the filename
        std::string search_term = clean_filename(filename);
        search_term = normalization::normalize_search_term_default(search_term);

        // Search for the game
        auto response = request("/search/autocomplete/" + url_encode(search_term));
        if (response.empty() || !get_bool(response, "success")) {
            return nullptr;
        }

        if (!response.contains("data") || !response["data"].is_array() || response["data"].empty()) {
            return nullptr;
        }

        // Build name to game map
        std::map<std::string, nlohmann::json> games_by_name;
        std::vector<std::string> names;
        for (const auto& item : response["data"]) {
            std::string game_name = get_string(item, "name");
            if (!game_name.empty()) {
                games_by_name[game_name] = item;
                names.push_back(game_name);
            }
        }

        // Find best match
        auto best = matching::find_best_match_simple(search_term, names);
        if (!best.found()) {
            return nullptr;
        }

        const auto& game = games_by_name[best.match];
        int game_id = static_cast<int>(get_number(game, "id"));
        auto artwork = fetch_all_artwork(game_id);

        auto result = std::make_unique<GameResult>();
        result->provider = name();
        result->provider_id = game_id;
        result->provider_ids = {{"steamgriddb", game_id}};
        result->name = get_string(game, "name");
        result->artwork = artwork;
        result->match_score = best.score;
        result->match_type = "filename";

        double release_date = get_number(game, "release_date");
        if (release_date > 0) {
            result->metadata.release_year = static_cast<int>(release_date);
        }

        result->raw_response = game;
        return result;
    }

    void heartbeat() override {
        if (!config_.is_configured()) {
            throw AuthError(name(), "provider not configured");
        }

        // Try a simple search to check connectivity
        auto response = request("/search/autocomplete/test");
        // If we get here without throwing, the provider is available
    }

    void close() override {}

private:
    /// @brief Makes an authenticated request to the SteamGridDB API
    nlohmann::json request(const std::string& endpoint, const cpr::Parameters& params = {}) {
        std::string api_key = config_.get_credential("api_key");

        std::string url = kBaseURL + endpoint;
        cpr::Response r = cpr::Get(
            cpr::Url{url},
            cpr::Header{
                {"Accept", "application/json"},
                {"Authorization", "Bearer " + api_key}
            },
            params,
            cpr::Timeout{config_.timeout * 1000}
        );

        if (r.status_code == 401) {
            throw AuthError(name(), "invalid API key");
        }

        if (r.status_code == 429) {
            throw RateLimitError(name());
        }

        if (r.status_code != 200) {
            throw ConnectionError(name(), "HTTP " + std::to_string(r.status_code));
        }

        try {
            return nlohmann::json::parse(r.text);
        } catch (const nlohmann::json::parse_error& e) {
            throw ConnectionError(name(), "failed to parse JSON response");
        }
    }

    /// @brief Builds filter parameters for artwork requests
    cpr::Parameters build_filter_params() {
        cpr::Parameters params;

        // Content filters
        if (nsfw_) {
            params.Add({"nsfw", "any"});
        } else {
            params.Add({"nsfw", "false"});
        }

        if (humor_) {
            params.Add({"humor", "any"});
        } else {
            params.Add({"humor", "false"});
        }

        if (epilepsy_) {
            params.Add({"epilepsy", "any"});
        } else {
            params.Add({"epilepsy", "false"});
        }

        return params;
    }

    /// @brief Fetches grids (covers) for a game
    std::vector<nlohmann::json> fetch_grids(int game_id) {
        try {
            auto response = request("/grids/game/" + std::to_string(game_id), build_filter_params());
            if (!get_bool(response, "success")) {
                return {};
            }

            if (!response.contains("data") || !response["data"].is_array()) {
                return {};
            }

            std::vector<nlohmann::json> grids;
            for (const auto& item : response["data"]) {
                if (item.is_object()) {
                    grids.push_back(item);
                }
            }
            return grids;
        } catch (...) {
            return {};
        }
    }

    /// @brief Fetches heroes (banners/backgrounds) for a game
    std::vector<nlohmann::json> fetch_heroes(int game_id) {
        try {
            auto response = request("/heroes/game/" + std::to_string(game_id), build_filter_params());
            if (!get_bool(response, "success")) {
                return {};
            }

            if (!response.contains("data") || !response["data"].is_array()) {
                return {};
            }

            std::vector<nlohmann::json> heroes;
            for (const auto& item : response["data"]) {
                if (item.is_object()) {
                    heroes.push_back(item);
                }
            }
            return heroes;
        } catch (...) {
            return {};
        }
    }

    /// @brief Fetches logos for a game
    std::vector<nlohmann::json> fetch_logos(int game_id) {
        try {
            auto response = request("/logos/game/" + std::to_string(game_id), build_filter_params());
            if (!get_bool(response, "success")) {
                return {};
            }

            if (!response.contains("data") || !response["data"].is_array()) {
                return {};
            }

            std::vector<nlohmann::json> logos;
            for (const auto& item : response["data"]) {
                if (item.is_object()) {
                    logos.push_back(item);
                }
            }
            return logos;
        } catch (...) {
            return {};
        }
    }

    /// @brief Fetches icons for a game
    std::vector<nlohmann::json> fetch_icons(int game_id) {
        try {
            auto response = request("/icons/game/" + std::to_string(game_id), build_filter_params());
            if (!get_bool(response, "success")) {
                return {};
            }

            if (!response.contains("data") || !response["data"].is_array()) {
                return {};
            }

            std::vector<nlohmann::json> icons;
            for (const auto& item : response["data"]) {
                if (item.is_object()) {
                    icons.push_back(item);
                }
            }
            return icons;
        } catch (...) {
            return {};
        }
    }

    /// @brief Fetches all artwork types for a game
    Artwork fetch_all_artwork(int game_id) {
        Artwork artwork;

        // Fetch grids (covers)
        auto grids = fetch_grids(game_id);
        if (!grids.empty()) {
            artwork.cover_url = get_string(grids[0], "url");
        }

        // Fetch heroes (banners/backgrounds)
        auto heroes = fetch_heroes(game_id);
        if (!heroes.empty()) {
            artwork.background_url = get_string(heroes[0], "url");
            if (heroes.size() > 1) {
                artwork.banner_url = get_string(heroes[1], "url");
            }
        }

        // Fetch logos
        auto logos = fetch_logos(game_id);
        if (!logos.empty()) {
            artwork.logo_url = get_string(logos[0], "url");
        }

        // Fetch icons
        auto icons = fetch_icons(game_id);
        if (!icons.empty()) {
            artwork.icon_url = get_string(icons[0], "url");
        }

        return artwork;
    }

    ProviderConfig config_;
    std::shared_ptr<Cache> cache_;
    bool nsfw_;
    bool humor_;
    bool epilepsy_;
};

namespace {
[[maybe_unused]] ProviderRegistrar steamgriddb_registrar(
    "steamgriddb",
    [](const ProviderConfig& config, std::shared_ptr<Cache> cache) {
        return std::make_unique<SteamGridDBProvider>(config, std::move(cache));
    });
}

}  // namespace retro_metadata
